server/src/controllers/course.ts

```diff
@@ -106,12 +109,130 @@ export async function addCourse(req: Request, res: Response): Promise<void> {
   }
 }
 
+enum GradingType {
+  PassFail = 'PASSFAIL',
+  Numerical = 'NUMERICAL'
+}
+
+enum Period {
+  I = 'I',
+  II = 'II',
+  III = 'III',
+  IV = 'IV',
+  V = 'V'
+}
+
+enum TeachingMethod {
+  Lecture = 'LECTURE',
+  Exam = 'EXAM'
+}
+
+interface CourseInstanceAddRequest {
+  gradingType: GradingType;
+  startingPeriod: Period;
+  endingPeriod: Period;
+  teachingMethod: TeachingMethod;
+  responsibleTeacher: number;
+  startDate: Date;
+  endDate: Date;
+}
+
+const courseInstanceAddRequestSchema: yup.AnyObjectSchema = yup.object().shape({
+  gradingType: yup
+    .string()
+    .oneOf([GradingType.PassFail, GradingType.Numerical])
+    .required(),
+  startingPeriod: yup
+    .string()
+    .oneOf([Period.I, Period.II, Period.III, Period.IV, Period.V])
+    .required(),
+  endingPeriod: yup
+    .string()
+    .oneOf([Period.I, Period.II, Period.III, Period.IV, Period.V])
+    .required(),
+  teachingMethod: yup
+    .string()
+    .oneOf([TeachingMethod.Lecture, TeachingMethod.Exam])
+    .required(),
+  responsibleTeacher: yup
+    .number()
+    .required(),
+  startDate: yup
+    .date()
+    .required(),
+  endDate: yup
+    .date()
+    .required()
+});
+
+export async function addCourseInstance(req: Request, res: Response): Promise<Response> {
+  try {
+    const courseId: number = Number(req.params.courseId);
+
+    await courseInstanceAddRequestSchema.validate(req.body, { abortEarly: false });
+
+    const request: CourseInstanceAddRequest = req.body;
+
+    const course: Course | null = await models.Course.findOne({
+      where: {
+        id: courseId
+      },
+    });
+
+    if (course == null) {
+      throw new Error(`Course with ID ${courseId} does not exist`);
+    }
+
+    const teacher: User | null = await models.User.findOne({
+      where: {
+        id: request.responsibleTeacher
+      },
+    });
+
+    // TODO: Also check that user has correct role
```

Comment:
Could this be best done in a middleware?

Reply:
Actually, this check shouldn't be done in the first place since course roles are instance specific. My thought process for adding this comment was that we wouldn't accidentally want to add a student as a teacher for a course for example, but I'll remove this comment now.